File: python3/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1 for _ in range(len(nums))]
        prefix, postfix = 1, 1
        for i in range(0, len(nums), 1):
            res[i] = prefix
            prefix *= nums[i]
        for i in range(len(nums)-1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        return res



        # --- O(N) RUNTIME, O(N) SPACE COMPLEXITY ---
        # arr1, arr2 = nums.copy(), nums.copy()
        # arr1.insert(0, 1)
        # arr2.append(1)
        
        # for i in range(1, len(arr1), 1):
        #     arr1[i] = arr1[i] * arr1[i-1]
        
        # for i in range(len(arr2)-2, -1, -1):
        #     arr2[i] = arr2[i] * arr2[i+1]

        # res = []
        # for i in range(len(nums)):
        #     res.append(arr1[i] * arr2[i+1])
        # return res

File: golang/238.go
func productExceptSelf(nums []int) []int {
    prefix := []int{1}
    postfix := []int{1}
    res := []int{}

    for i := 0; i < len(nums) - 1; i++ {
        prefix = append(prefix, prefix[len(prefix) - 1] * nums[i])
    }
    for i := len(nums) - 1; i > 0; i-- {
        postfix = append(postfix, postfix[len(postfix) - 1] * nums[i])
    }
    for i := 0; i < len(nums); i++ {
        res = append(res, prefix[i] * postfix[len(nums) - 1 - i])
    }
    return res
}
